### PR TITLE
Deploy to the Cloud guides (Kubernetes/Docker and Cloud Foundry)

### PR DESCRIPTION
**Pull request by alexwishart**
These replace the current docker and cloud deploy guides.

### CODE REVIEW

File: guides/deploy/starter_deploy_cloudfoundry.html [hunk 1]
```diff
@@ -37,7 +37,7 @@ <h1 class="header-title">KITURA</h1>
       <h1>DEPLOY YOUR APP USING CLOUD FOUNDRY</h1>
     </div>
 
-<p>This guide will show you how to deploy the Kitura application you created by following our <a href="https://www.kitura.io/guides/gettingstarted">Getting Started</a> guide to a <a href="#cloud-foundry">Cloud Foundry</a> based cloud environment. The guide assumes you already have an environment to deploy to. If not, you can <a href="https://console.bluemix.net/registration/">sign up</a> to IBM Cloud for free and use it to host your Cloud Foundry apps.</p>
+<p>This guide will show you how to deploy the Kitura application you created by following our <a href="guides/gettingstarted.html">Getting Started</a> guide to a <a href="#cloud-foundry">Cloud Foundry</a> based cloud environment. The guide assumes you already have an environment to deploy to. If not, you can <a href="https://console.bluemix.net/registration/">sign up</a> to IBM Cloud for free and use it to host your Cloud Foundry apps.</p>
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_cloudfoundry.html]: These relative links don't work, I think this one needs to be ../gettingstarted.md but you'll need to test it.

File: guides/deploy/starter_deploy_kubernetes.html [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+<p>To install the chart into your cluster, run the following from the <code>/chart/&lt;your app name&gt;</code> directory:</p>
+<div><pre><code class="sourceCode bash">helm install --name myapp .</code></pre></div>
+<p>You may have to wait a few minutes for it to be available, but your application should now be deployed to your Kubernetes cluster. To access your application, you have to find the external IP of the worker node it has been deployed to. This process varies, so check with your cloud provider on how to do so. The port that your application had been deployed on is described by the <code>NodePort</code> which you can find by running the following command on your cluster:</p>
+<div><pre><code class="sourceCode bash">kubectl describe service &lt;your app name&gt;-service</code></pre></div>
+<p>Congratulations! You have now successfully deployed your Kitura application using Kubernetes! If you visit <code>&lt;external IP&gt;:&lt;NodePort&gt;</code> you should se the Kitura splash page.</p>
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_kubernetes.html]: "you should se" -> "you should see"

Also this code tag looks like straight text <code>&lt;external IP&gt;:&lt;NodePort&gt;</code> I suggest maybe changing it to <highlight>

File: guides/deploy/starter_deploy_kubernetes.html [hunk 3]
```diff
@@ -37,7 +37,7 @@ <h1 class="header-title">KITURA</h1>
...
+<p>This guide will show you how to deploy the Kitura application you created by following our <a href="guides/gettingstarted.html">Getting Started</a> guide to a <a href="https://kubernetes.io/">Kubernetes</a> based environment. The guide assumes you already have an environment to deploy to. If not, you can <a href="https://console.bluemix.net/registration/">sign up</a> to IBM Cloud and <a href="https://console.bluemix.net/docs/containers/container_index.html#container_index">create a Kubernetes cluster</a> for free, or <a href="https://kubernetes.io/docs/setup/minikube/">set one up locally using minikube</a>.</p>
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_kubernetes.html]: Don't we need a capital M for minikube? The Getting Started link needs fixing here too.

File: guides/deploy/starter_deploy_cloudfoundry.html [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        <img width="480px" src="../../assets/Kitura.svg" alt="Kitura Logo">
+      <h1>DEPLOY YOUR APP USING CLOUD FOUNDRY</h1>
+    </div>
+
+<p>This guide will show you how to deploy the Kitura application you created by following our <a href="https://www.kitura.io/guides/gettingstarted">Getting Started</a> guide to a <a href="#cloud-foundry">Cloud Foundry</a> based cloud environment. The guide assumes you already have an environment to deploy to. If not, you can <a href="https://console.bluemix.net/registration/">sign up</a> to IBM Cloud for free and use it to host your Cloud Foundry apps.</p>
+<h4 >Starter project contents</h4>
+<p>Your generated Kitura project already contains a manifest.yml file which you can use to deploy your application. This file is optional, but allows you to maintain consistency across deployments. If necessary you can adjust or add attributes, a list of which can be found <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#optional-attributes">here</a>. You’ll use the Cloud Foundry command line interface, which you can install using the instructions <a href="https://github.com/cloudfoundry/cli#downloads">here</a>.</p>
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_cloudfoundry.html]: How about...

Your generated Kitura project contains a manifest.yml file which you can use to deploy your application. Although you can deploy applications without a manifest, manifests provide consistency and reproducibility across deployments. You can also specify <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#optional-attributes"> additional application attributes</a> within your manifest.

File: guides/deploy/starter_deploy_cloudfoundry.html [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  <a href="http://slack.kitura.io/">
+    <img width="80px" src="../../assets/slack-icon.png" alt="Slack icon">
+    <h2>NEED HELP?</h2>
+    <h2>MESSAGE US ON SLACK.</h2>
+  </a>
+</section>
+</body>
+<footer>
+
+<nav class="footer-nav">
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_cloudfoundry.html]: David tells me that the footer is in the wrong place in this file.

File: guides/deploy/starter_deploy_kubernetes.html [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    </div>
+
+<p>This guide will show you how to deploy the Kitura application you created by following our <a href="https://www.kitura.io/guides/gettingstarted">Getting Started</a> guide to a <a href="https://kubernetes.io/">Kubernetes</a> based environment. The guide assumes you already have an environment to deploy to. If not, you can <a href="https://console.bluemix.net/registration/">sign up</a> to IBM Cloud and <a href="https://console.bluemix.net/docs/containers/container_index.html#container_index">create a Kubernetes cluster</a> for free, or <a href="https://kubernetes.io/docs/setup/minikube/">set one up locally using minikube</a>.</p>
+<h4 id="starter-project-contents">Starter project contents</h4>
+<p>Your generated Kitura project already contains Docker related files and a <a href="https://docs.helm.sh/developing_charts/">Helm chart</a> used to describe the set of Kubernetes resources it requires.</p>
+<p>To deploy your application to Kubernetes, we first need to build it into a Docker container for Kubernetes to execute. Your project includes the following files for Docker support: - .dockerignore - Dockerfile - Dockerfile-tools</p>
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_kubernetes.html]: : - .dockerignore - Dockerfile - Dockerfile-tools
this format isn't readable when rendered. Perhaps a list would be better?

File: guides/deploy/starter_deploy_kubernetes.html [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+<p>The .dockerignore file contains the files/directories that should not be included in the built Docker image. By default this file contains the Dockerfile and Dockerfile-tools. It can be modified as required.</p>
+<p><code>Dockerfile</code> defines the specification of the default Docker image for running the application. This image can be used to <em>run</em> the application.</p>
+<p><code>Dockerfile-tools</code> is a Docker specification file similar to a Dockerfile, except it includes the tools required for compiling the application. This image can be used to <em>compile</em> the application.</p>
+<h4 id="build-and-push-your-docker-image">Build and push your Docker image</h4>
+<p>You need to build the Docker image that will run your Kitura application and will deployed to your Kubernetes cluster. The following commands should all be run from the directory containing your Dockerfiles.</p>
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_kubernetes.html]: will deployed -> will be deployed

File: guides/deploy/starter_deploy_kubernetes.html [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+<p>The Helm chart provided configures a number of default resource options, such as memory allowance and replica policies. You can make changes to this depending on your or your cloud provider’s requirements.</p>
+<p>To install the chart into your cluster, run the following from the <code>/chart/&lt;your app name&gt;</code> directory:</p>
+<div><pre><code class="sourceCode bash">helm install --name myapp .</code></pre></div>
+<p>You may have to wait a few minutes for it to be available, but your application should now be deployed to your Kubernetes cluster. To access your application, you have to find the external IP of the worker node it has been deployed to. This process varies, so check with your cloud provider on how to do so. The port that your application had been deployed on is described by the <code>NodePort</code> which you can find by running the following command on your cluster:</p>
```
Comment (helenmasters) [inline guides/deploy/starter_deploy_kubernetes.html]: had been deployed on -> has been deployed on

File: guides/deploy/starter_deploy_cloudfoundry.html (hunk 1 above)
Comment (ddunn2) [reply guides/deploy/starter_deploy_cloudfoundry.html]: Correct, when using relative links you need to step back up using `../`. 
In this case you'd want to do `../gettingstarted.html` (we don't have md files currently). 
But as Helen suggested, I'd test that link actually works :)

File: guides/deploy/starter_deploy_kubernetes.html (hunk 2 above)
Comment (ddunn2) [reply guides/deploy/starter_deploy_kubernetes.html]: For now lets just highlight them, you can do this by wrapping the text in `<span>` tags like so: 
`<span class="highlight">Text you want highlighting</span>`

When I move onto refactoring the all of the guides styling/layout, I'll be addressing the this styling as well and will update it to whatever we decide.

File: guides/deploy/starter_deploy_cloudfoundry.html (hunk 5 above)
Comment (ddunn2) [reply guides/deploy/starter_deploy_cloudfoundry.html]: Correct, the footer in this guide is incorrect. It's sitting inside the white section of the guide rather than the foot of the page. 

After looking at the code, it's not that the footer is in the wrong place it seems you've missed a closing `</section>` tag. 

If you add a `</section>` tag just before the `</body>` tag it should resolve this issue :).

**Comment by ddunn2** [comment]
@alexwishart The issue is the guide didn't seem to work for me, I didn't get an external ip for the last step for example.

I'm not sure if it's an issue with the guide directly, but the guide is hard to follow if you're new to Kubernetes, Helm or cloud providers. 
There's a lot of places I was expected to navigate away from the guide and have to learn how various other technologies work together before I could return to the guide to attempt to complete it. 

I mentioned this briefly before, but if we're suggesting Docker Hub then maybe we should include instructions for that.

**Comment by ddunn2** [comment]
Just to clear up my last comment:

- Can you follow this guide with a `kitura init` project and it works? For me there's a few extra steps.

- Would be nice to include a section about pushing to the registry. This is such an important part of the process I feel an example should be included. Even if we just pick one, and let users discover the rest for themselves. 

- A nice to have would be to explain what the commands are actually doing, but not essential at this point. 

Hopefully this clears up the points I was making in my previous comment :)

**Comment by alexwishart** [comment]
@ddunn2 My idea about the guide was that it was going to have to make some assumptions. The only way we can make sure we don't pidgeonhole people into using IBM Cloud (which is something @ianpartridge was keen on) is to keep the guide a bit vague. I agree, it makes things like pushing to the registry or finding the external ip for the kube service difficult, but the guide is about deploying the `kitura init` app and I think going too in depth into setting up a registry or your kube cluster could distract from that.
I've been able to run through the guide from a `kitura init` project.. I've taken the `helm init` command to come under the assumption that you already have a kube cluster set up, maybe this is something that should be expressed more explicitly at the top of the guide? :)
Overall it was a pretty difficult guide to bring together because of the vagueness so I appreciate any comments. In my opinion the best way forward would be to create/link to additional guides for the instructions you feel are missing?

**Comment by ddunn2** [comment]
I feel a guide should literally be guiding a user through doing something step by step, if we're only providing half of the instructions and assuming people should learn the rest they may lose interest pretty quickly, it can be a time consuming process to pull together all the different commands you need to actually complete this guide especially if you're new to it all. 

If we're explicit in the guide that you can use whatever cloud service you'd like and then mention that in this guide we'll be using IBM cloud as an example? I understand we don't want to pigeonhole people but if we're going to provide instructions it should be around IBM Cloud :)

Vague guides are difficult to write, and difficult to follow. Personally I'd prefer the guide to be explicit, a one stop place for anyone wanting to deploy an app using Kubernetes, whether they're wanting to use IBM Cloud or not. 

Maybe something like: 

If you already have a deployment environment set up, you can skip to the [deploy]() section. 
If not then the following instructions will take you through how to setup an environment using IBM Cloud...

**Comment by christiancompton** [comment]
If you do feel that content specifically around Kitura and IBM Cloud is necessary but not appropriate for Kitura.io, we can put that into the Swift programming guide, and you could link to that.  

@jpkessle and christx83 have joined our team and have been hard at work on making programming guides. Just let us know!

**Comment by ddunn2** [comment]
Thanks @christiancompton 

If there is a nice succinct guide we could link to around setting up an IBM Cloud environment for Kube that'd be useful. 

In the meantime we could just include the 'high level' commands here, that is the bare minimum for users to be able to complete this guide starting from nothing. 
That is: 
Link to CLI installation/account creation (or just provide the commands)
Create a cluster
Create a namespace 
Show how to get the public/external ip of cluster

With those four things anyone could complete this guide :)

**Comment by ddunn2** [comment]
Just going to merge this for now, we can update this at a later date :)
